python/lsst/db/utils.py

```diff
@@ -21,49 +19,350 @@
 # see <http://www.lsstcorp.org/LegalNotices/>.
 
 """
-This module contains small utilities / helpers that perform common tasks related to
-the Db wrapper.
-
+This module contains utilities / helpers that perform common tasks, in particular these that are specific to different drivers.
 
 @author  Jacek Becla, SLAC
 """
 
 
 # standard library
-import ConfigParser
-import os.path
-import sys
-
-# local
-from lsst.db.db import Db
-
-
-def readCredentialFile(fName, logger):
-    """
-    Reads all supported key/value pairs from fName and return a dictionary
-    containing these key/value pairs translated to names accepted by connect()
-    as needed). Hint, to get a subset, do something like:
-    dict = readCredentialFile(fN)
-    (hst, prt, usr, pwd) = [dict[k] for k in ('host', 'port', 'user', 'passwd')].
-    This function only reads from the [mysql] section, e.g., it is not full
-    equivalent to how mysql command like utility which obtains the value from the
-    last occurrence of k in section [mysql] or [client] in the file.
-    """
-    ret = {}
-    if fName.startswith('~'):
-        fName = os.path.expanduser(fName)
-    if not os.path.isfile(fName):
-        raise Exception("Required file '%s' not found" % fName)
-    cnf = ConfigParser.ConfigParser()
-    cnf.read(fName)
-
-    theSection = "mysql"
-    if not cnf.has_section(theSection):
-        raise Exception("Missing section '%s' in '%s'" % (theSection, fName))
-    for o in Db.optionToConnectArgMap:
-        if cnf.has_option(theSection, o):
-            theKey = Db.optionToConnectArgMap.get(o, o)
-            ret[theKey] = cnf.get(theSection, o)
-    logger.info("fetched %s from '%s' (password not shown)" % (
-            str(["%s:%s" % (x, ret[x]) for x in ret if not x == "passwd"]), fName))
-    return ret
+import logging as log
+import re
+import os
+
+# third party
+import sqlalchemy
+from sqlalchemy.exc import DBAPIError, InvalidRequestError, NoSuchModuleError, \
+                           NoSuchTableError, OperationalError, ProgrammingError
+from sqlalchemy.sql import text
+from sqlalchemy.inspection import inspect
+
+
+
+# MySQL errors that we are catching
+# Names and numbers from include/mysql/mysql.h
+class MySqlErr:
+    ER_DB_DROP_EXISTS = 1008
+    ER_NO_DB_ERROR = 1046
+    ER_BAD_DB_ERROR = 1049
+    ER_TABLE_EXISTS_ERROR = 1050
+    ER_BAD_TABLE_ERROR = 1051
+
+
+class NoSuchDatabaseError(InvalidRequestError):
+    """Database does not exist."""
+
+class InvalidDatabaseNameError(InvalidRequestError):
+    """Invalid database name."""
+
+class DatabaseExistsError(InvalidRequestError):
+    """Database already exists."""
+
+class TableExistsError(InvalidRequestError):
+    """Table already exists."""
+
+#### Database-related functions ####################################################
+def createDb(conn, dbName, mayExist=False):
+    """
+    Create database <dbName>.
+
+    @param conn        Database connection.
```

Comment:
We are doing wrappers for sqlalchemy, I thought it may be worth to keep the same approach and allow engine/connection interchangeably where it makes sense.